views/default/page/layouts/widgets/add_panel.php

```diff
@@ -1,69 +1,87 @@
- * @uses $vars['container']   Container to optional limit widgets for. Defaults to page_owner_entity
  */
 
-$widgets = elgg_extract('widgets', $vars);
+if (!elgg_is_xhr()) {
+	echo elgg_echo('ajax:not_is_xhr');
+	return;
+}
+
 $context = elgg_extract('context', $vars);
-$exact = elgg_extract('exact_match', $vars, false);
-$container = elgg_extract('container', $vars, elgg_get_page_owner_entity());
+$exact = elgg_extract('exact_match', $vars);
+$owner_guid = (int) elgg_extract('owner_guid', $vars);
+$owner = get_entity($owner_guid);
+elgg_set_page_owner_guid($owner_guid);
+
+$context_stack = elgg_extract('context_stack', $vars);
+if (!empty($context_stack)) {
+	elgg_set_context_stack($context_stack);
+}
 
+$widgets = elgg_get_widgets($owner->guid, $context);
 $widget_types = elgg_get_widget_types([
 	'context' => $context,
 	'exact' => $exact,
-	'container' => $container,
+	'container' => $owner,
 ]);
 uasort($widget_types, function($a, $b) {
-	return strcmp($a->name,$b->name);
+	return strcmp($a->name, $b->name);
 });
 
-$current_handlers = array();
+$current_handlers = [];
 foreach ($widgets as $column_widgets) {
 	foreach ($column_widgets as $widget) {
 		$current_handlers[] = $widget->handler;
 	}
 }
 
-?>
-<div class="elgg-widgets-add-panel hidden clearfix" id="widgets-add-panel">
-	<p>
-		<?php echo elgg_echo('widgets:add:description'); ?>
-	</p>
-	<ul>
-<?php
-		foreach ($widget_types as $handler => $widget_type) {
-			// check if widget added and only one instance allowed
-			if ($widget_type->multiple == false && in_array($handler, $current_handlers)) {
-				$class = 'elgg-state-unavailable';
-				$tooltip = elgg_echo('widget:unavailable');
-			} else {
-				$class = 'elgg-state-available';
-				$tooltip = $widget_type->description;
-			}
+$result = elgg_autop(elgg_echo('widgets:add:description'));
+
+$list_items = '';
+foreach ($widget_types as $handler => $widget_type) {
+	$class = [];
+	// check if widget added and only one instance allowed
+	if (!$widget_type->multiple && in_array($handler, $current_handlers)) {
+		$class[] = 'elgg-state-unavailable';
+	} else {
+		$class[] = 'elgg-state-available';
+	}
+
+	$class[] = $widget_type->multiple ? 'elgg-widget-multiple' : 'elgg-widget-single';
+	
+	$item_content = '<div class="elgg-widgets-add-actions float-alt">';
+	if (!$widget_type->multiple) {
+		$item_content .= elgg_format_element('span', ['class' => 'elgg-quiet'], elgg_echo('widget:unavailable'));
+	}
+	$add_link = elgg_http_add_url_query_elements('action/widgets/add', [
+		'handler' => $handler,
+		'page_owner_guid' => $owner_guid,
+		'context' => $context,
+		'show_access' => elgg_extract('show_access', $vars),
+		'default_widgets' => 0,
+	]);
+					
+	$item_content .= elgg_view('output/url', [
+		'class' => 'elgg-button elgg-button-submit',
+		'text' => elgg_echo('add'),
+		'href' => $add_link,
+		'is_action' => true,
+	]);
+	$item_content .= '</div>';
+	
+	$item_content .= "<h3>{$widget_type->name}</h3>";
```

Comment:
changed to h4 (as module heading is already h3).... but should module heading not be a h2?... other discussion